lib/ruby_lsp/requests/hover.rb

```diff
@@ -0,0 +1,109 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "ruby_lsp/requests/support/rails_document_client"
+
+module RubyLsp
+  module Requests
+    # ![Hover demo](../../misc/rails_document_link_hover.gif)
+    #
+    # The [hover request](https://microsoft.github.io/language-server-protocol/specification#textDocument_hover)
+    # renders a clickable link to the code's official documentation.
+    # It currently only supports Rails' documentation: when hovering over Rails DSLs/constants under certain paths,
+    # like `before_save :callback` in `models/post.rb`, it generates a link to `before_save`'s API documentation.
+    #
+    # # Example
+    #
+    # ```ruby
+    # class Post < ApplicationRecord
+    #   before_save :do_something # when hovering on before_save, the link will be rendered
+    # end
+    # ```
+    class Hover < BaseRequest
+      extend T::Sig
+
+      sig { params(document: Document, uri: String, position: Document::PositionShape).void }
+      def initialize(document, uri, position)
+        super(document)
+
+        @file_dir = T.let(Pathname.new(uri).dirname.to_s, String)
+        @hover = T.let(nil, T.nilable(LanguageServer::Protocol::Interface::Hover))
+        position = Document::Scanner.new(document.source).find_position(position)
+        @target = T.let(find(T.must(@document.tree), position), T.nilable(SyntaxTree::Node))
+      end
+
+      sig { override.returns(T.nilable(LanguageServer::Protocol::Interface::Hover)) }
+      def run
+        return if @target.nil?
+
+        visit(@document.tree)
+        @hover
+      end
+
+      sig { override.params(node: SyntaxTree::Command).void }
+      def visit_command(node)
+        return unless node == @target
+
+        message = node.message
+        generate_rails_document_link_hover(message.value, message)
+      end
+
+      sig { override.params(node: SyntaxTree::FCall).void }
+      def visit_fcall(node)
+        return unless node == @target
+
+        message = node.value
+        generate_rails_document_link_hover(message.value, message)
+      end
+
+      sig { override.params(node: SyntaxTree::ConstPathRef).void }
+      def visit_const_path_ref(node)
+        return unless node == @target
+
+        constant_name = full_constant_name(node)
+        generate_rails_document_link_hover(constant_name, node)
+      end
+
+      private
+
+      sig { params(name: String, node: SyntaxTree::Node).void }
+      def generate_rails_document_link_hover(name, node)
+        urls = Support::RailsDocumentClient.generate_rails_document_urls(name, @file_dir)
+
+        unless urls.empty?
+          values = urls.map do |url|
+            "[Rails Document: `#{url[:name]}`](#{url[:url]})"
```

Comment:
NIT: Can we extract the `Rails Document` part so that it's not a part of the links? Also, what do you think about joining the different entries with pipes instead of new lines?

Example result
```
Rails documentation: SomeClass#method | OtherClass#other_method
```

Reply:
> Also, what do you think about joining the different entries with pipes instead of new lines?

Given that Rails components' namespaces usually aren't that short, I feel using `|` will make the line too long. Based on what I see, I think newline is still better.

![Multiple links](https://user-images.githubusercontent.com/5079556/193134223-6443d957-c9c4-4adc-9a92-b9cb2c6da5f0.png)